### PR TITLE
Fix anchor invalid redirection to search

### LINKED ISSUE #41933: Clicking on rustdoc sidebar item results in 'undefined' page
https://doc.rust-lang.org/nightly/std/index.html

Click on 'Primitive Types' in the sidebar

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #41933.

r? @rust-lang/docs

### CODE REVIEW

File: src/librustdoc/html/static/main.js [hunk 1]
```diff
@@ -935,37 +935,39 @@
...
+                        document.title = previousTitle;
+                        // When browsing forward to search results the previous
+                        // search will be repeated, so the currentResults are
+                        // cleared to ensure the search is successful.
+                        currentResults = null;
+                        // Synchronize search bar with query string state and
+                        // perform the search. This will empty the bar if there's
+                        // nothing there, which lets you really go back to a
+                        // previous state with nothing in the bar.
+                        document.getElementsByClassName('search-input')[0].value = params.search;
```
Comment (ollie27) [inline src/librustdoc/html/static/main.js]: If `params.search` is `undefined` this should set the value to `""` not `"undefined"` like it currently does.

File: src/librustdoc/html/static/main.js [hunk 2]
```diff
@@ -935,37 +935,39 @@
...
-                    // visibility must be reset.
-                    if (!params.search) {
-                        var main = document.getElementById('main');
-                        if (hasClass(main, 'content')) {
-                            removeClass(main, 'hidden');
-                        }
-                        var search = document.getElementById('search');
-                        if (hasClass(main, 'content')) {
-                            addClass(main, 'hidden');
+                    if (location.href.indexOf(".html?search=") !== -1) {
```
Comment (ollie27) [inline src/librustdoc/html/static/main.js]: This would break clearing the search results so should be removed.

**Review by ollie27 — CHANGES_REQUESTED** [review]

File: src/librustdoc/html/static/main.js (hunk 2 above)
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js]: It seems it's still working. So not sure if I tested what you had in mind.

File: src/librustdoc/html/static/main.js [hunk 3]
```diff
@@ -935,37 +935,39 @@
...
+                        // visibility must be reset.
+                        if (!params.search) {
+                            var main = document.getElementById('main');
+                            if (hasClass(main, 'content')) {
+                                removeClass(main, 'hidden');
+                            }
+                            var search = document.getElementById('search');
+                            if (hasClass(main, 'content')) {
+                                addClass(main, 'hidden');
+                            }
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js]: Excellent point!

File: src/librustdoc/html/static/main.js (hunk 1 above)
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js]: 👍

File: src/librustdoc/html/static/main.js (hunk 2 above)
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js]: Oh my bad, I need to test on a non local path.

File: src/librustdoc/html/static/main.js (hunk 3 above)
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js]: This was the issue actually. I thought the variable was limited to the scope, big mistake!

**Comment by GuillaumeGomez** [comment]
Updated.

**Review by ollie27 — APPROVED** [review]

**Review by frewsxcv — APPROVED** [review]

**Pull request merged**